genie-core/src/main/java/com/netflix/genie/core/jpa/repositories/JpaClusterRepository.java

```diff
@@ -16,13 +16,87 @@
+            + "        FROM"
+            + "          clusters c join"
+            + "          clusters_tags ct on c.id = ct.cluster_id join"
+            + "          tags t on ct.tag_id = t.id"
+            + "        WHERE"
+            + "          t.tag IN (:clusterTags) AND"
+            + "          c.status = 'UP'"
+            + "        GROUP BY"
+            + "          c.id"
+            + "        HAVING"
+            + "          COUNT(c.id) = :clusterTagsCount"
+            + "      ) AS selected_clusters join"
+            + "      clusters_commands cc ON selected_clusters.id = cc.cluster_id join"
+            + "      ("
+            + "        SELECT"
+            + "          c.id as id"
+            + "        FROM"
+            + "          commands c join"
+            + "          commands_tags ct on c.id = ct.command_id join"
+            + "          tags t on ct.tag_id = t.id"
+            + "        WHERE"
+            + "          t.tag IN (:commandTags) AND"
+            + "          c.status = 'ACTIVE'"
+            + "        GROUP BY"
+            + "          c.id"
+            + "        HAVING"
+            + "          COUNT(c.id) = :commandTagsCount"
+            + "      ) AS selected_commands ON selected_commands.id = cc.command_id"
+            + "    GROUP BY cc.cluster_id"
+            + "  ) as cluster_id_order join"
+            + "  clusters_commands cc on"
+            + "    cluster_id_order.cluster_id = cc.cluster_id AND"
+            + "    cc.command_order = cluster_id_order.command_order join"
+            + "  commands c on cc.command_id = c.id;";
+
+    /**
+     * Find the cluster and command ids for the given criterion tags. The data is returned as a list of
+     * Long, Long results where the first Long is the cluster id and the second is the command id that should be used
+     * if that cluster is selected. The command id should be the highest priority command that matches the command
+     * criterion which has been associated with the given cluster by administrators.
+     *
+     * @param clusterTags      The tags the cluster must match to be selected (conjunction). e.g. 'tag1','tag2','tag3'
+     * @param clusterTagsCount The number of cluster tags (size of clusterTags)
+     * @param commandTags      The tags the command associated with the cluster must match to be selected. (conjunction)
+     *                         e.g. 'tag1','tag2'
+     * @param commandTagsSize  The number of command tags (size of commandTags)
+     * @return The id pairs found or empty list if no matches
+     */
+    @Query(value = CLUSTER_COMMAND_QUERY, nativeQuery = true)
+    List<Object[]> findClustersAndCommandsForCriterion(
+        @Param("clusterTags") final Set<String> clusterTags,
+        @Param("clusterTagsCount") final int clusterTagsCount,
+        @Param("commandTags") final Set<String> commandTags,
+        @Param("commandTagsCount") final int commandTagsSize
```

Comment:
That would be nice however I didn't really find a quick way to do that with the parameter filled methods. Could probably look around for something but didn't think it was a big enough deal to go into [SpEL](https://docs.spring.io/spring/docs/3.0.x/spring-framework-reference/html/expressions.html) or something. That may be smart enough to pull out the sizes of the collections but it would really lose compile time checks.